lambda/css-api/src/services/keycloak-service.ts

```diff
@@ -0,0 +1,422 @@
+import axios, { AxiosError, AxiosInstance } from 'axios';
+import { getKeycloakCredentials } from '../util';
+import jwt from 'jsonwebtoken';
+import { RolePayload } from '../types';
+import createHttpError from 'http-errors';
+import { inject, singleton } from 'tsyringe';
+
+interface KeycloakTokenResponse {
+  access_token: string;
+  refresh_token: string;
+  expires_in: number;
+  refresh_expires_in: number;
+  token_type: string;
+  'not-before-policy': number;
+  session_state: string;
+  scope: string;
+}
+
+export class KeycloakService {
+  private realm: string = 'standard';
+  private httpClient: AxiosInstance;
+  private environment: string;
+  private accessToken: string | null = null;
+  private refreshToken: string | null = null;
+  private refreshing: boolean = false;
+
+  constructor() {}
+
+  setEnvironment(environment: string) {
+    this.environment = environment;
+    const { keycloakUrl } = getKeycloakCredentials(environment);
+    this.httpClient = axios.create({
+      baseURL: `${keycloakUrl}`,
+    });
+    this.httpClient.interceptors.response.use(
+      (response) => {
+        return response;
+      },
+      (error: AxiosError) => {
+        console.error('keycloak request path: ', error?.request?.path);
+        console.error('keycloak request headers: ', error?.request?.headers);
+        console.error('keycloak response status: ', error?.response?.status);
+        console.error('keycloak response data: ', error?.response?.data);
+        console.error('keycloak response status message: ', error?.response?.statusText);
+        throw error;
+      },
+    );
+  }
+
+  getEnvironment() {
+    return this.environment;
+  }
+
+  isTokenValid(accessToken: string) {
+    return (jwt.decode(accessToken) as any).exp > Date.now().valueOf() / 1000 + 30;
+  }
+
+  async getAccessToken() {
+    if (this.accessToken && this.isTokenValid(this.accessToken)) {
+      return this.accessToken;
+    }
+    if (
+      this.accessToken &&
+      !this.isTokenValid(this.accessToken) &&
+      this.refreshToken &&
+      this.isTokenValid(this.refreshToken)
+    ) {
+      // refreshing access token 30 seconds earlier
+      if (this.refreshing) {
+        return this.accessToken;
```

Comment:
Wouldn't this return expired access tokens sometimes? e.g getAccessToken is called when the token is just about to expire two times in close succession. On the second call this.refreshing would be true, but the access token could have expired while the refresh was still in flight.

I think to make a lock we could have this.refreshing be a promise that resolves after setting the refreshed token, then this could await the promise and return the access token after

Reply:
@jlangy I am refreshing the token halfway through the access token lifespan so we have 30 seconds to complete the refresh token flow, which I believe is a lot of time as it takes less than a second to complete this flow in actual. The locking mechanism wouldn't benefit as per my testing with K6. K6 initiates multiple threads and if one thread has a lock for refreshing token, the other threads still keep running using expired access tokens leading to failures. Unless we refresh the token way earlier, it doesn't make it different that what I have implemented.